File: docs/examples/nts-server.js
#!/usr/bin/env node
// Minimal Network Time Security server (RFC 8915), Node.js, no dependencies.
//
// It speaks exactly the exchange documented in docs/time-server.md, so
// Lockasaur (or any RFC 8915 client) can use it as a trusted time source:
//
//   1. NTS-KE over TLS 1.3 (TCP): negotiate NTPv4 + AEAD_AES_SIV_CMAC_256,
//      export the c2s/s2c keys, hand out cookies.
//   2. NTS-protected NTPv4 (UDP): authenticate the request, answer with the
//      current time signed under the s2c key.
//
// Cookies here are the two session keys sealed under a per-run master key, so
// the stateless UDP side can recover them without shared storage. A production
// server would rotate that master key on a schedule and persist it.
//
// Run:
//   openssl req -x509 -newkey rsa:2048 -nodes -keyout key.pem -out cert.pem \
//     -days 365 -subj "/CN=localhost" \
//     -addext "subjectAltName=DNS:localhost,IP:127.0.0.1"
//   node nts-server.js
//
// Lockasaur verifies the certificate against the OS trust store, so pointing
// the app at a real deployment needs a publicly trusted certificate (for
// example from Let's Encrypt) and the KE port reachable on TCP 4460.

import tls from "node:tls";
import dgram from "node:dgram";
import crypto from "node:crypto";
import fs from "node:fs";

const KE_PORT = Number(process.env.NTS_KE_PORT || 4460);
const NTP_PORT = Number(process.env.NTS_NTP_PORT || 4461);
const CERT = process.env.NTS_CERT || "cert.pem";
const KEY = process.env.NTS_KEY || "key.pem";

const MASTER_KEY = crypto.randomBytes(32);

// ---------------------------------------------------------------------------
// AES-SIV-CMAC-256 (RFC 5297) on AES-128 primitives. Node exposes no SIV
// cipher, so S2V/CMAC and CTR are built by hand. The 32-byte key splits into
// K1 (CMAC) and K2 (CTR).
// ---------------------------------------------------------------------------

const BLOCK = 16;
const ZERO = Buffer.alloc(BLOCK);

function dbl(b) {
    const out = Buffer.alloc(BLOCK);
    let carry = 0;
    for (let i = BLOCK - 1; i >= 0; i--) {
        out[i] = ((b[i] << 1) | carry) & 0xff;
        carry = b[i] >>> 7;
    }
    if (carry) out[BLOCK - 1] ^= 0x87;
    return out;
}

function xor(a, b) {
    const out = Buffer.alloc(BLOCK);
    for (let i = 0; i < BLOCK; i++) out[i] = a[i] ^ b[i];
    return out;
}

function ecb(key16, block) {
    const c = crypto.createCipheriv("aes-128-ecb", key16, null);
    c.setAutoPadding(false);
    return Buffer.concat([c.update(block), c.final()]);
}

function cmac(key16, data) {
    const k1 = dbl(ecb(key16, ZERO));
    const k2 = dbl(k1);
    const blocks = Math.max(1, Math.ceil(data.length / BLOCK));
    const whole = data.length > 0 && data.length % BLOCK === 0;
    let last;
    if (whole) {
        last = xor(data.slice((blocks - 1) * BLOCK), k1);
    } else {
        const padded = Buffer.alloc(BLOCK);
        data.copy(padded, 0, (blocks - 1) * BLOCK);
        padded[data.length % BLOCK] = 0x80;
        last = xor(padded, k2);
    }
    let x = ZERO;
    for (let i = 0; i < blocks - 1; i++) x = ecb(key16, xor(x, data.slice(i * BLOCK, (i + 1) * BLOCK)));
    return ecb(key16, xor(x, last));
}

function s2v(key16, strings) {
    if (strings.length === 0) {
        const one = Buffer.alloc(BLOCK);
        one[BLOCK - 1] = 0x01;
        return cmac(key16, one);
    }
    let d = cmac(key16, ZERO);
    for (let i = 0; i < strings.length - 1; i++) d = xor(dbl(d), cmac(key16, strings[i]));
    const last = strings[strings.length - 1];
    let t;
    if (last.length >= BLOCK) {
        t = Buffer.from(last);
        const off = last.length - BLOCK;
        for (let i = 0; i < BLOCK; i++) t[off + i] ^= d[i];
    } else {
        const padded = Buffer.alloc(BLOCK);
        last.copy(padded);
        padded[last.length] = 0x80;
        t = xor(dbl(d), padded);
    }
    return cmac(key16, t);
}

// The CTR IV clears bit 31 of each of the last two 32-bit words (RFC 5297).
function ctrIv(siv) {
    const q = Buffer.from(siv);
    q[8] &= 0x7f;
    q[12] &= 0x7f;
    return q;
}

function ctr(key16, iv, data) {
    const c = crypto.createCipheriv("aes-128-ctr", key16, iv);
    return Buffer.concat([c.update(data), c.final()]);
}

function sivSeal(key32, adList, plaintext) {
    const siv = s2v(key32.slice(0, BLOCK), [...adList, plaintext]);
    return Buffer.concat([siv, ctr(key32.slice(BLOCK), ctrIv(siv), plaintext)]);
}

function sivOpen(key32, adList, sealed) {
    if (sealed.length < BLOCK) return null;
    const siv = sealed.slice(0, BLOCK);
    const plaintext = ctr(key32.slice(BLOCK), ctrIv(siv), sealed.slice(BLOCK));
    const expected = s2v(key32.slice(0, BLOCK), [...adList, plaintext]);
    return crypto.timingSafeEqual(expected, siv) ? plaintext : null;
}

// ---------------------------------------------------------------------------
// Cookies: AES-256-GCM(masterKey) over c2s||s2c. Layout nonce(12) | tag(16) | ct.
// ---------------------------------------------------------------------------

function sealCookie(c2s, s2c) {
    const nonce = crypto.randomBytes(12);
    const c = crypto.createCipheriv("aes-256-gcm", MASTER_KEY, nonce);
    const ct = Buffer.concat([c.update(Buffer.concat([c2s, s2c])), c.final()]);
    return Buffer.concat([nonce, c.getAuthTag(), ct]);
}

function openCookie(cookie) {
    if (cookie.length !== 12 + 16 + 64) return null;
    try {
        const d = crypto.createDecipheriv("aes-256-gcm", MASTER_KEY, cookie.slice(0, 12));
        d.setAuthTag(cookie.slice(12, 28));
        const keys = Buffer.concat([d.update(cookie.slice(28)), d.final()]);
        return { c2s: keys.slice(0, 32), s2c: keys.slice(32, 64) };
    } catch {
        return null;
    }
}

// ---------------------------------------------------------------------------
// NTS-KE (TLS 1.3, ALPN ntske/1).
// ---------------------------------------------------------------------------

const CRITICAL = 0x8000;
const REC_EOM = 0;
const REC_NEXT_PROTOCOL = 1;
const REC_AEAD = 4;
const REC_COOKIE = 5;
const REC_NTP_PORT = 7;
const PROTO_NTPV4 = 0;
const AEAD_AES_SIV_CMAC_256 = 15;

const EXPORTER_LABEL = "EXPORTER-network-time-security";
const KEY_LEN = 32;

function record(type, body) {
    const buf = Buffer.alloc(4 + body.length);
    buf.writeUInt16BE(type, 0);
    buf.writeUInt16BE(body.length, 2);
    body.copy(buf, 4);
    return buf;
}

function u16(value) {
    const b = Buffer.alloc(2);
    b.writeUInt16BE(value, 0);
    return b;
}

// Both keys derive from the TLS session: 5-byte context is protocol | AEAD |
// direction (0x00 client-to-server, 0x01 server-to-client).
function exportKey(socket, direction) {
    const ctx = Buffer.alloc(5);
    ctx.writeUInt16BE(PROTO_NTPV4, 0);
    ctx.writeUInt16BE(AEAD_AES_SIV_CMAC_256, 2);
    ctx.writeUInt8(direction, 4);
    return socket.exportKeyingMaterial(KEY_LEN, EXPORTER_LABEL, ctx);
}

// True once the accumulated request contains an End of Message record.
function requestComplete(buf) {
    let offset = 0;
    for (;;) {
        if (offset + 4 > buf.length) return false;
        const type = buf.readUInt16BE(offset) & ~CRITICAL;
        const bodyLen = buf.readUInt16BE(offset + 2);
        if (offset + 4 + bodyLen > buf.length) return false;
        offset += 4 + bodyLen;
        if (type === REC_EOM) return true;
    }
}

function buildKeResponse(cookies) {
    const parts = [
        record(CRITICAL | REC_NEXT_PROTOCOL, u16(PROTO_NTPV4)),
        record(REC_AEAD, u16(AEAD_AES_SIV_CMAC_256)),
        record(REC_NTP_PORT, u16(NTP_PORT))
    ];
    for (const cookie of cookies) parts.push(record(REC_COOKIE, cookie));
    parts.push(record(CRITICAL | REC_EOM, Buffer.alloc(0)));
    return Buffer.concat(parts);
}

const keServer = tls.createServer(
    {
        cert: fs.readFileSync(CERT),
        key: fs.readFileSync(KEY),
        minVersion: "TLSv1.3",
        ALPNProtocols: ["ntske/1"]
    },
    (socket) => {
        socket.on("error", () => {});
        if (socket.alpnProtocol !== "ntske/1") {
            socket.destroy();
            return;
        }
        const chunks = [];
        socket.on("data", (chunk) => {
            chunks.push(chunk);
            const buf = Buffer.concat(chunks);
            if (buf.length > 4096 || !requestComplete(buf)) return;
            const c2s = exportKey(socket, 0x00);
            const s2c = exportKey(socket, 0x01);
            const cookies = [sealCookie(c2s, s2c), sealCookie(c2s, s2c)];
            socket.end(buildKeResponse(cookies));
        });
    }
);

// ---------------------------------------------------------------------------
// NTS-protected NTPv4 (UDP).
// ---------------------------------------------------------------------------

const NTP_HEADER_LEN = 48;
const ORIGIN_OFFSET = 24;
const XMT_OFFSET = 40;
const EF_UNIQUE_ID = 0x0104;
const EF_COOKIE = 0x0204;
const EF_AUTHENTICATOR = 0x0404;
const NTP_UNIX_OFFSET = 2208988800;

function extensionField(type, body, minBody = 0) {
    const padded = Math.max(body.length, minBody);
    const total = 4 + padded + ((4 - ((4 + padded) % 4)) % 4);
    const buf = Buffer.alloc(total);
    buf.writeUInt16BE(type, 0);
    buf.writeUInt16BE(total, 2);
    body.copy(buf, 4);
    return buf;
}

function ntpTimestamp() {
    const ms = Date.now();
    const seconds = Math.floor(ms / 1000) + NTP_UNIX_OFFSET;
    const fraction = Math.floor(((ms % 1000) / 1000) * 4294967296);
    const buf = Buffer.alloc(8);
    buf.writeUInt32BE(seconds >>> 0, 0);
    buf.writeUInt32BE(fraction >>> 0, 4);
    return buf;
}

// Collect the extension fields the request carries, stopping at the
// authenticator (everything before it is the authenticated associated data).
function readRequest(buf) {
    const out = { authStart: -1 };
    let offset = NTP_HEADER_LEN;
    while (offset + 4 <= buf.length) {
        const type = buf.readUInt16BE(offset);
        const total = buf.readUInt16BE(offset + 2);
        if (total < 4 || total % 4 !== 0 || offset + total > buf.length) break;
        const body = buf.slice(offset + 4, offset + total);
        if (type === EF_UNIQUE_ID) {
            out.uniqueId = body.slice(0, 32);
        } else if (type === EF_COOKIE) {
            out.cookie = body;
        } else if (type === EF_AUTHENTICATOR) {
            const nonceLen = body.readUInt16BE(0);
            const cipherLen = body.readUInt16BE(2);
            out.authStart = offset;
            out.nonce = body.slice(4, 4 + nonceLen);
            out.sealed = body.slice(4 + nonceLen, 4 + nonceLen + cipherLen);
            break;
        }
        offset += total;
    }
    return out;
}

function handleNtp(req) {
    if (req.length < NTP_HEADER_LEN) return null;
    const parsed = readRequest(req);
    if (!parsed.uniqueId || !parsed.cookie || parsed.authStart < 0) return null;

    const keys = openCookie(parsed.cookie);
    if (!keys) return null;

    const associated = req.slice(0, parsed.authStart);
    if (sivOpen(keys.c2s, [associated, parsed.nonce], parsed.sealed) === null) return null;

    const header = Buffer.alloc(NTP_HEADER_LEN);
    header.writeUInt8(0x24, 0); // LI 0, VN 4, mode 4 (server)
    header.writeUInt8(1, 1); // stratum 1 (primary reference)
    header.writeInt8(-23, 3); // precision, roughly one microsecond
    header.write("LOCL", 12, "ascii"); // reference identifier
    req.copy(header, ORIGIN_OFFSET, XMT_OFFSET, XMT_OFFSET + 8); // echo client transmit as origin
    ntpTimestamp().copy(header, 32); // receive
    const now = ntpTimestamp();
    now.copy(header, 16); // reference
    now.copy(header, XMT_OFFSET); // transmit: this is the time verdict the client reads

    const idField = extensionField(EF_UNIQUE_ID, parsed.uniqueId);
    const preceding = Buffer.concat([header, idField]);
    const nonce = crypto.randomBytes(16);
    const sealed = sivSeal(keys.s2c, [preceding, nonce], Buffer.alloc(0));
    const authBody = Buffer.alloc(4 + nonce.length + sealed.length);
    authBody.writeUInt16BE(nonce.length, 0);
    authBody.writeUInt16BE(sealed.length, 2);
    nonce.copy(authBody, 4);
    sealed.copy(authBody, 4 + nonce.length);
    return Buffer.concat([preceding, extensionField(EF_AUTHENTICATOR, authBody, 24)]);
}

const ntpServer = dgram.createSocket("udp4");
ntpServer.on("message", (msg, rinfo) => {
    const response = handleNtp(msg);
    if (response) ntpServer.send(response, rinfo.port, rinfo.address);
});

keServer.listen(KE_PORT, () => console.log(`NTS-KE on tcp/${KE_PORT}`));
ntpServer.bind(NTP_PORT, () => console.log(`NTS-NTP on udp/${NTP_PORT}`));

File: docs/examples/nts-server.py
#!/usr/bin/env python3
# Minimal Network Time Security server (RFC 8915), Python.
#
# Speaks the exact exchange documented in docs/time-server.md, so Lockasaur (or
# any RFC 8915 client) can use it as a trusted time source:
#
#   1. NTS-KE over TLS 1.3 (TCP): negotiate NTPv4 + AEAD_AES_SIV_CMAC_256,
#      export the c2s/s2c keys, hand out cookies.
#   2. NTS-protected NTPv4 (UDP): authenticate the request, answer with the
#      current time signed under the s2c key.
#
# Python's standard-library ssl module cannot export TLS keying material, so the
# KE side uses pyOpenSSL; AES-SIV uses pycryptodome:
#
#   pip install pyOpenSSL pycryptodome
#
# Cookies here are the two session keys sealed under a per-run master key, so the
# stateless UDP side can recover them without shared storage. A production server
# would rotate that master key on a schedule and persist it.
#
# Run:
#   openssl req -x509 -newkey rsa:2048 -nodes -keyout key.pem -out cert.pem \
#     -days 365 -subj "/CN=localhost" \
#     -addext "subjectAltName=DNS:localhost,IP:127.0.0.1"
#   python3 nts-server.py
#
# Lockasaur verifies the certificate against the OS trust store, so a real
# deployment needs a publicly trusted certificate (for example from Let's
# Encrypt) and the KE port reachable on TCP 4460.

import os
import socket
import struct
import threading
import time

from OpenSSL import SSL
from Crypto.Cipher import AES

KE_PORT = int(os.environ.get("NTS_KE_PORT", "4460"))
NTP_PORT = int(os.environ.get("NTS_NTP_PORT", "4461"))
CERT = os.environ.get("NTS_CERT", "cert.pem")
KEY = os.environ.get("NTS_KEY", "key.pem")
HOST = "0.0.0.0"

MASTER_KEY = os.urandom(32)

EXPORTER_LABEL = b"EXPORTER-network-time-security"

# --------------------------------------------------------------------------
# AEAD_AES_SIV_CMAC_256 (RFC 5297). pycryptodome's MODE_SIV takes one update()
# per S2V component, with the nonce as the last associated-data item; the tag it
# returns is the synthetic IV. Sealed layout is SIV(16) || ciphertext.
# --------------------------------------------------------------------------

def siv_seal(key, ad_list, plaintext):
    cipher = AES.new(key, AES.MODE_SIV)
    for ad in ad_list:
        cipher.update(ad)
    ciphertext, tag = cipher.encrypt_and_digest(plaintext)
    return tag + ciphertext


def siv_open(key, ad_list, sealed):
    if len(sealed) < 16:
        return None
    cipher = AES.new(key, AES.MODE_SIV)
    for ad in ad_list:
        cipher.update(ad)
    try:
        return cipher.decrypt_and_verify(sealed[16:], sealed[:16])
    except ValueError:
        return None


# Cookies: AES-256-GCM(master key) over c2s||s2c. Layout nonce(12) | tag(16) | ct.

def seal_cookie(c2s, s2c):
    nonce = os.urandom(12)
    cipher = AES.new(MASTER_KEY, AES.MODE_GCM, nonce=nonce)
    ciphertext, tag = cipher.encrypt_and_digest(c2s + s2c)
    return nonce + tag + ciphertext


def open_cookie(cookie):
    if len(cookie) != 12 + 16 + 64:
        return None
    try:
        cipher = AES.new(MASTER_KEY, AES.MODE_GCM, nonce=cookie[:12])
        keys = cipher.decrypt_and_verify(cookie[28:], cookie[12:28])
        return keys[:32], keys[32:64]
    except ValueError:
        return None


# --------------------------------------------------------------------------
# NTS-KE (TLS 1.3, ALPN ntske/1).
# --------------------------------------------------------------------------

CRITICAL = 0x8000
REC_EOM = 0
REC_NEXT_PROTOCOL = 1
REC_AEAD = 4
REC_COOKIE = 5
REC_NTP_PORT = 7
PROTO_NTPV4 = 0
AEAD_AES_SIV_CMAC_256 = 15


def record(rec_type, body):
    return struct.pack(">HH", rec_type, len(body)) + body


def u16(value):
    return struct.pack(">H", value)


# The 5-byte exporter context is protocol | AEAD | direction (0x00 c2s, 0x01 s2c).
def context(direction):
    return struct.pack(">HHB", PROTO_NTPV4, AEAD_AES_SIV_CMAC_256, direction)


def request_complete(buf):
    offset = 0
    while True:
        if offset + 4 > len(buf):
            return False
        rec_type = struct.unpack_from(">H", buf, offset)[0] & ~CRITICAL
        body_len = struct.unpack_from(">H", buf, offset + 2)[0]
        if offset + 4 + body_len > len(buf):
            return False
        offset += 4 + body_len
        if rec_type == REC_EOM:
            return True


def build_ke_response(cookies):
    parts = [
        record(CRITICAL | REC_NEXT_PROTOCOL, u16(PROTO_NTPV4)),
        record(REC_AEAD, u16(AEAD_AES_SIV_CMAC_256)),
        record(REC_NTP_PORT, u16(NTP_PORT)),
    ]
    for cookie in cookies:
        parts.append(record(REC_COOKIE, cookie))
    parts.append(record(CRITICAL | REC_EOM, b""))
    return b"".join(parts)


def make_tls_context():
    ctx = SSL.Context(SSL.TLS_SERVER_METHOD)
    ctx.set_min_proto_version(SSL.TLS1_3_VERSION)
    ctx.use_privatekey_file(KEY)
    ctx.use_certificate_file(CERT)

    def select_alpn(conn, protocols):
        return b"ntske/1" if b"ntske/1" in protocols else SSL.NO_OVERLAPPING_PROTOCOLS

    ctx.set_alpn_select_callback(select_alpn)
    return ctx


def handle_ke(ctx, raw):
    conn = SSL.Connection(ctx, raw)
    conn.set_accept_state()
    try:
        conn.do_handshake()
        if conn.get_alpn_proto_negotiated() != b"ntske/1":
            return
        buf = b""
        while not request_complete(buf):
            data = conn.recv(4096)
            if not data or len(buf) > 4096:
                return
            buf += data
        c2s = conn.export_keying_material(EXPORTER_LABEL, 32, context(0x00))
        s2c = conn.export_keying_material(EXPORTER_LABEL, 32, context(0x01))
        conn.sendall(build_ke_response([seal_cookie(c2s, s2c), seal_cookie(c2s, s2c)]))
        conn.shutdown()
    except SSL.Error:
        pass
    finally:
        raw.close()


# --------------------------------------------------------------------------
# NTS-protected NTPv4 (UDP).
# --------------------------------------------------------------------------

NTP_HEADER_LEN = 48
EF_UNIQUE_ID = 0x0104
EF_COOKIE = 0x0204
EF_AUTHENTICATOR = 0x0404
NTP_UNIX_OFFSET = 2208988800


def extension_field(ef_type, body, min_body=0):
    padded = max(len(body), min_body)
    total = 4 + padded
    total += (4 - (total % 4)) % 4
    buf = bytearray(total)
    struct.pack_into(">HH", buf, 0, ef_type, total)
    buf[4:4 + len(body)] = body
    return bytes(buf)


def ntp_timestamp():
    now = time.time()
    seconds = int(now) + NTP_UNIX_OFFSET
    fraction = int((now - int(now)) * (1 << 32))
    return struct.pack(">II", seconds & 0xFFFFFFFF, fraction & 0xFFFFFFFF)


# Read the request extension fields, stopping at the authenticator (everything
# before it is the authenticated associated data).
def read_request(buf):
    result = {"auth_start": -1}
    offset = NTP_HEADER_LEN
    while offset + 4 <= len(buf):
        ef_type, total = struct.unpack_from(">HH", buf, offset)
        if total < 4 or total % 4 != 0 or offset + total > len(buf):
            break
        body = buf[offset + 4:offset + total]
        if ef_type == EF_UNIQUE_ID:
            result["unique_id"] = bytes(body[:32])
        elif ef_type == EF_COOKIE:
            result["cookie"] = bytes(body)
        elif ef_type == EF_AUTHENTICATOR:
            nonce_len, cipher_len = struct.unpack_from(">HH", body, 0)
            result["auth_start"] = offset
            result["nonce"] = bytes(body[4:4 + nonce_len])
            result["sealed"] = bytes(body[4 + nonce_len:4 + nonce_len + cipher_len])
            break
        offset += total
    return result


def handle_ntp(request):
    if len(request) < NTP_HEADER_LEN:
        return None
    parsed = read_request(request)
    if "unique_id" not in parsed or "cookie" not in parsed or parsed["auth_start"] < 0:
        return None
    keys = open_cookie(parsed["cookie"])
    if keys is None:
        return None
    c2s, s2c = keys
    associated = request[:parsed["auth_start"]]
    if siv_open(c2s, [associated, parsed["nonce"]], parsed["sealed"]) is None:
        return None

    header = bytearray(NTP_HEADER_LEN)
    header[0] = 0x24  # LI 0, VN 4, mode 4 (server)
    header[1] = 1  # stratum 1 (primary reference)
    struct.pack_into(">b", header, 3, -23)  # precision, roughly one microsecond
    header[12:16] = b"LOCL"  # reference identifier
    header[24:32] = request[40:48]  # origin echoes the client transmit timestamp
    header[32:40] = ntp_timestamp()  # receive
    now = ntp_timestamp()
    header[16:24] = now  # reference
    header[40:48] = now  # transmit: the time verdict the client reads

    id_field = extension_field(EF_UNIQUE_ID, parsed["unique_id"])
    preceding = bytes(header) + id_field
    nonce = os.urandom(16)
    sealed = siv_seal(s2c, [preceding, nonce], b"")
    auth_body = struct.pack(">HH", len(nonce), len(sealed)) + nonce + sealed
    return preceding + extension_field(EF_AUTHENTICATOR, auth_body, 24)


def ke_server():
    ctx = make_tls_context()
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind((HOST, KE_PORT))
    sock.listen(16)
    print("NTS-KE on tcp/%d" % KE_PORT, flush=True)
    while True:
        raw, _ = sock.accept()
        threading.Thread(target=handle_ke, args=(ctx, raw), daemon=True).start()


def ntp_server():
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind((HOST, NTP_PORT))
    print("NTS-NTP on udp/%d" % NTP_PORT, flush=True)
    while True:
        data, addr = sock.recvfrom(65535)
        try:
            response = handle_ntp(data)
        except Exception:
            response = None
        if response:
            sock.sendto(response, addr)


if __name__ == "__main__":
    threading.Thread(target=ntp_server, daemon=True).start()
    ke_server()
